Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -149,4 +165,405 @@ theorem rootsOfUnity_eq_torsion [NumberField K] :
 
 end torsion
 
+namespace dirichletUnitTheorem
+
+-- This section is devoted to the proof of Dirichlet's unit theorem
+-- We define a group morphism from `(𝓞 K)ˣ` to `{w : InfinitePlace K // w ≠ w₀} → ℝ` where `w₀`
+-- is a distinguished (arbitrary) infinite place, prove that its kernel is the torsion subgroup
+-- (see `logEmbedding_eq_zero_iff`) and that its image, called `unitLattice`, is a full
+-- `ℤ`-lattice. It follows that is a free `ℤ`-module (see `unitLattice_moduleFree `) of
+-- rank `card (InfinitePlaces K) - 1` (see `unitLattice_rank`).
+-- To prove that the `unitLattice` is a full `ℤ`-lattice, we need to prove that it is discrete
+-- (see `unitLattice_inter_ball_finite`) and that it spans the full space over `ℝ`
+-- (see ` unitLattice_span_eq_top`); this is the main part of the proof, see the section
+-- `span_top` below for more details.
+
+open scoped Classical
+
+variable [NumberField K]
+
+variable {K}
+
+/-- The distinguished infinite place. -/
+def w₀ : InfinitePlace K := (inferInstance : Nonempty (InfinitePlace K)).some
+
+variable (K)
+
+/-- The logarithmic embedding of the units (seen as an `Additive` group). -/
+def logEmbedding : Additive ((𝓞 K)ˣ) →+ ({w : InfinitePlace K // w ≠ w₀} → ℝ) :=
+{ toFun := fun x w => mult w.val * Real.log (w.val (Additive.toMul x))
+  map_zero' := by simp; rfl
+  map_add' := by
+    intro _ _
+    simp [ne_eq, toMul_add, map_mul, map_eq_zero, coe_ne_zero, Real.log_mul, mul_add]
+    rfl }
+
+variable {K}
+
+@[simp]
+theorem logEmbedding_component (x : (𝓞 K)ˣ) (w : {w : InfinitePlace K // w ≠ w₀}) :
```

Comment:
This can be called `logEmbedding_apply` (but I am not sure). I tried to auto generate it using `simps`, but unfortunately it produces a syntactically different lemma (even if it is definitionally the same).

Reply:
I'd rather keep the name as it is even though I agree it is not great since, for me, in this context `apply` means applying a function to its argument and that would not be the case here.